[GeoMechanicsApplication] Fixed setting "DOMAIN_SIZE" by the `GeoMechanicalSolver`

avdg81 opened the pull request:
**📝 Description**

During a [previous cleanup action](https://github.com/KratosMultiphysics/Kratos/pull/12950), there was accidentally a bug introduced. When a new model part was created, we no longer added the `DOMAIN_SIZE` to the `ProcessInfo` instance of the model part. As a result, the master-slave constraints process failed (due to the absence of `DOMAIN_SIZE`). The geomechanical solver has been corrected to overcome this issue. A Python integration test has been added to demonstrate the correctness of the fix.

avdg81 committed:
Fixed setting "DOMAIN_SIZE" by the `GeoMechanicalSolver`

Also added an integration test for master-slave constraints that reproduced the original issue, where the master-slave constraints were not properly set up due to a missing value for "DOMAIN_SIZE" at the `ProcessInfo` instance.

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py:
```diff
@@ -34,13 +34,13 @@ def __init__(self, model, custom_settings):
             if domain_size < 0:
                 raise ValueError('Please specify a "domain_size" >= 0')
             self.main_model_part.ProcessInfo.SetValue(KratosMultiphysics.DOMAIN_SIZE, domain_size)
+        else:
+            self.main_model_part = self.model[model_part_name]
```
yes, this should fix my mistake.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_master_slave_constraints.py:
```diff
@@ -0,0 +1,47 @@
+        ):
+            self.assertAlmostEqual(
+                actual_displacement[0], expected_x_displacement, places=3
+            )
```
I wonder if this is something that is already available in our test helper.

See are_iterables_almost_equal

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_master_slave_constraints.py:
```diff
@@ -0,0 +1,47 @@
+        for actual_reaction_force, expected_x_reaction_force in zip(
+            actual_reaction_forces, expected_x_reaction_forces
+        ):
+            self.assertAlmostEqual(actual_reaction_force[0], expected_x_reaction_force)
```
same hopefully available comparison.

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_master_slave_constraints/ProjectParameters.json:
```diff
@@ -0,0 +1,172 @@
+                "model_part_name"        : "PorousDomain.porous_computational_model_part",
+                "master_model_part_name" : "PorousDomain.MasterSide",
+                "slave_model_part_name"  : "PorousDomain.SlaveSide",
+                "variable_names"         : ["DISPLACEMENT"],
```
This is a list, it could include the WATER_PRESSURE.

avdg81 replied:
Good catch! By adding `"WATER_PRESSURE"` here, we also "tie" the water pressures (in addition to the displacements). That is a more complete description, although it won't have an impact on the test itself (since we don't calculate the groundwater pressure field).

rfaasse commented on applications/GeoMechanicsApplication/tests/test_master_slave_constraints/README.md:
```diff
@@ -0,0 +1,11 @@
+
+To keep the model as simple as possible, a linear elastic material model has been applied.  Also, the groundwater pressure field has been fixed at a value of 0.0 throughout the entire domain.
+
+The goal of this test is to show that master-slave constraints are correctly applied.  We verify that by comparing the actual displacements in $`x`$ direction along the loaded side with the analytical solution.  Furthermore, we compare the reaction forces in $`x`$ direction along the supported side with the analytical solution.
```
It might be helpful to show the analytical formula here

avdg81 replied:
I have added analytical formulae to calculate the expected displacements in $`x`$ direction at the loaded side, as well as the reaction force component in $`x`$ direction at the supported side.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_master_slave_constraints.py:
```diff
@@ -0,0 +1,47 @@
+        ):
+            self.assertAlmostEqual(
+                actual_displacement[0], expected_x_displacement, places=3
+            )
```
Next to checking the loaded nodes, it might be nice to also check that the displacements of the 'tied' nodes are equal

avdg81 replied:
I have added an assertion for this. We now check whether the displacements in $`x`$ direction at the "tied" node pairs compare equal.

rfaasse reviewed (dismissed):
Nice catch and clear well-tested fix! I have two minor suggestions, but I think this shouldn't take long to get back to master 👍

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_master_slave_constraints/README.md:
```diff
@@ -0,0 +1,11 @@
+
+To keep the model as simple as possible, a linear elastic material model has been applied.  Also, the groundwater pressure field has been fixed at a value of 0.0 throughout the entire domain.
+
+The goal of this test is to show that master-slave constraints are correctly applied.  We verify that by comparing the actual displacements in $`x`$ direction along the loaded side with the analytical solution.  Furthermore, we compare the reaction forces in $`x`$ direction along the supported side with the analytical solution.
```
The formatting seems to consistently have 2 spaces after each sentence closed by a "."
Why?
Maybe the experts Marjan and Richard can help with a picture.

avdg81 replied:
I use this convention to easily find the end of a sentence in the source text. It won't have any effect on the generated text. Also, two spaces helps to quickly jump over other uses of dots (e.g. in abbreviations).

I will ask one of Marjan or Richard to help me (quickly) make a picture in Inkscape.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_master_slave_constraints/ProjectParameters.json:
```diff
@@ -0,0 +1,172 @@
+            "model_part_name": "PorousDomain.FixedNode",
+            "variable_name":   "DISPLACEMENT",
+            "active":          [true,true,true],
+            "is_fixed":        [true,true,true],
```
Probably we can keep the z components false. ( also in the next vector constraint table process )

avdg81 replied:
Yes, that makes things simpler. Done.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_master_slave_constraints/ProjectParameters.json:
```diff
@@ -0,0 +1,172 @@
+        "strategy_type":                      "linear",
+        "convergence_criterion":              "residual_criterion",
+        "displacement_relative_tolerance":    1.0E-4,
+        "displacement_absolute_tolerance":    1.0E-9,
```
maybe leave out the not used criteria values ( displacement and water_pressure )
Ideally the .json file contains only the necessary minimum of information.

avdg81 replied:
Yes, this again makes the input files simpler. I have removed the redundant tolerance values for displacements and water pressures.

WPK4FEM reviewed (commented):
Dear Anne,
Thank you for correcting my mistake. Very good to have the check in our integration test suite.
Some small remarks but nothing blocking.
Regards, Wijtze Pieter

avdg81 committed:
Processed Richard's review suggestions

- Added the analytical formulae for calculating the expected displacements in $`x`$ direction at the loaded side, and the reaction force components in $`x`$ direction at the supported side.
- Added one more assertion to the test: the "tied" nodes should have equal displacements in $`x`$ direction.

avdg81 committed:
Processed Wijtze Pieter's review suggestions

- Use a helper function for comparing two lists of numbers.
- Also include the `"WATER_PRESSURE"` in the master-slave constraints (since we use them with U-Pw elements).
- Deactivated the $`z`$ component of the prescribed displacements (since we have a plane strain model).

avdg81 committed:
Forgot to add changes of Wijtze Pieter's final suggestion

avdg81 committed:
The GeoMechanicsApp now extends rather than replaces `Register`

We have found that during test runs Kratos applications deregister themselves **and** the Kratos Core. However, when a test run continues, the Kratos Core is not always properly registered again, which lead to an unexpected test failure due to `LinearMasterSlaveConstraint` not being a registered Kratos component. By letting the GeoMechanicsApplication extend rather than replace the base class implementation of `Register`, we have overcome this problem. It seems that repeatedly calling the base class's `Register` member doesn't do any harm.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_master_slave_constraints/README.md:
Could you still add the image to the README.md?

avdg81 replied:
Done.

rfaasse reviewed (commented):
Thank you for incorporating the changes, and the hard work, this turned out to be more difficult than expected in the end! To me this is ready to go when the schematic is added to the readme 👍

avdg81 committed:
Further improved the test documentation

- Wijtze Pieter had a few more suggestions for the text.

avdg81 committed:
Further improved the test documentation

- Added the figure of the adopted mesh that had been created by Richard, but which I had forgotten to add to the documentation.
- Wijtze Pieter had a few more suggestions for the text.

avdg81 committed:
Made the figure somewhat bigger (to improve readability)
